src/components/pages/profile/account/SetMFA.vue

```diff
@@ -0,0 +1,371 @@
+  selectedStepIndex: 0,
+  steps: [
+    {
+      name: t(`accountPage.mfa.scan.title`),
+      validated: false,
+      error: false,
+    },
+    {
+      name: t(`accountPage.mfa.enterCode.title`),
+      validated: false,
+      error: false,
+    },
+    {
+      name: t(`accountPage.mfa.result.title`),
+      validated: false,
+      error: false,
+    },
+  ] as Step[],
+});
+
+const props = defineProps<{
+  contactId: string;
+}>();
+
+/** Authenticator app url to scan (otpauth://totp/...) */
+const totpUrl = ref<string | undefined>(undefined);
+
+/** Information about the app totp is set up for */
+const totpIdentity = ref<TotpIdentity>({
+  issuer: undefined,
+  label: undefined,
+});
+
+/** Secret used to generate totp */
+const totpSecret = ref(new Secret());
+
+/** User token used to verify totp  */
+const userToken = ref('');
+
+/** Is the user token valid? */
+const userTokenValid = ref(false);
+
+/** TOTP instance */
+let totp: TOTP | null = null;
+
+const modalTitle = computed(() => t(`accountPage.mfa.modalTitle`));
+const codeLabel = computed(() => t(`accountPage.mfa.codeInput.label`));
+
+/** Called when the modal is closed */
+const onCloseMFAModal = () => {
+  closeMFAModal();
+  resetState();
+};
+
+/** Close the modal */
+const closeMFAModal = () => {
+  showMFASettingsModal.value = false;
+};
+
+/** Toggle the modal */
+const toggleMFAModal = () => {
+  showMFASettingsModal.value = !showMFASettingsModal.value;
+};
+
+const saveMFA = async () => {
+  closeMFAModal();
+  const result = await createContactMfa(props.contactId, {
+    secret: totpSecret.value.base32,
+    token: userToken.value,
+    type: ContactMfaType.TOTP,
+  });
+  isEnabled.value = true;
+  resetState();
+  return result;
+};
+
+/**
+ * Disable MFA for the contact
+ * TODO: Show confirmation dialog
+ */
+const disableMFA = async () => {
+  await deleteContactMfa(props.contactId);
+  isEnabled.value = false;
+  resetState();
+};
+
+/** Called when the slider changes */
+const onSlideChange = (details: AppSliderSlideEventDetails) => {
+  syncStepperWithSlider(details);
+  validateStep(details);
+};
+
+/** Reset / init the state of the component */
+const resetState = () => {
+  appSliderCo.value?.toSlide(0);
+  appStepper.value.selectedStepIndex = 0;
+  appStepper.value.steps.forEach((step) => {
+    step.validated = false;
+    step.error = false;
+  });
+  userToken.value = '';
+  userTokenValid.value = false;
+};
+
+/** Sync the stepper with the slider */
+const syncStepperWithSlider = (details: AppSliderSlideEventDetails) => {
+  if (appStepper.value.selectedStepIndex === details.slideNumber) {
+    return;
+  }
+  appStepper.value.selectedStepIndex = details.slideNumber;
+
+  validatePreviousSteps();
+};
+
+/** Validate the current step */
+const validateStep = (details: AppSliderSlideEventDetails) => {
+  if (details.slideNumber === 2) {
+    validateTOTOToken();
+  }
+};
+
+/** Validate all previous steps */
+const validatePreviousSteps = () => {
```

Comment:
There *should* be a much nicer way to do validation here, but I know this is also how the `CalloutStepper` does it.

In most places we use the vuelidate library and it has some really nice features that would almost certainly eliminate the need for this sort of thing, but I'm not sure how right now. Let's leave it like this for now then when I refactor `CalloutStepper` I might have a little look at this too.

Reply:
I will take a closer look, in my case I actually only do a real validation in the last slide, in the previous ones it is more of a pseudo validation.